openmls/src/framing/message_out.rs

```diff
@@ -0,0 +1,187 @@
+//! MLS Message (Output)
+//!
+//! This module defines the [`MlsMessageOut`] structs which implements the
+//! `MLSMessage` struct as defined by the MLS specification, but is used
+//! exclusively as output of the [`MlsGroup`] API. [`MlsMessageIn`] also
+//! implements `MLSMessage`, but for inputs.
+//!
+//! The [`MlsMessageOut`] struct is meant to be serialized upon its return from
+//! a function of the `MlsGroup` API so that it can be sent to the DS.
+use tls_codec::Serialize;
+
+use super::*;
+
+use crate::{key_packages::KeyPackage, versions::ProtocolVersion};
+
+/// An [`MlsMessageOut`] is typically returned from an [`MlsGroup`] function and
+/// meant to be serialized and sent to the DS.
+#[derive(Debug, Clone, PartialEq, TlsSerialize, TlsSize)]
+#[cfg_attr(feature = "test-utils", derive(TlsDeserialize))]
+pub struct MlsMessageOut {
+    pub(crate) version: ProtocolVersion,
+    pub(crate) body: MlsMessageOutBody,
+}
+
+/// MLSMessage (Body)
+///
+/// Note: Because [MlsMessageOutBody] already discriminates between
+/// `public_message`, `private_message`, etc., we don't use the
+/// `wire_format` field. This prevents inconsistent assignments
+/// where `wire_format` contradicts the variant given in `body`.
+///
+/// ```c
+/// // draft-ietf-mls-protocol-17
+/// struct {
+///     // ... continued from [MlsMessage] ...
+///
+///     WireFormat wire_format;
+///     select (MLSMessage.wire_format) {
+///         case mls_plaintext:
+///             PublicMessage plaintext;
+///         case mls_ciphertext:
+///             PrivateMessage ciphertext;
+///         case mls_welcome:
+///             Welcome welcome;
+///         case mls_group_info:
+///             GroupInfo group_info;
+///         case mls_key_package:
+///             KeyPackage key_package;
+///     }
+/// } MLSMessage;
+/// ```
+#[derive(Debug, PartialEq, Clone, TlsSerialize, TlsSize)]
+#[cfg_attr(feature = "test-utils", derive(TlsDeserialize))]
+#[repr(u8)]
+pub(crate) enum MlsMessageOutBody {
+    /// Plaintext message
+    #[tls_codec(discriminant = 1)]
```

Comment:
It would be nice if we wouldn't have to define the wire format value twice. Maybe for a follow up.

Reply:
That would be nice, but I think we have to if we want to include different message types for `In` and `Out` message. This is already the case for `GroupInfo` vs. `VerifiableGroupInfo` and there are going to be more values like that in the future.